File: Programmers/jsh/12주차/여행경로.py
'''
그래프 문제이므로 배열또는 사전자료형 사용
여기선 원소가 string이므로 dictionary 사용
양방향 그래프가 아닐수도 있기때문에 모든경우를 탐색해야함(DFS)
순서에따라 경유하지 못하는 공항이 생길수있음
티켓사용시 리스트에서 제거해야하므로 pop하고 만약 끝까지 탐색후 정답이아니라면 
다시 돌아와서 pop된원소를 다시 push해주고 다음 원소를 pop후 다시탐색(27줄)
-------------------------------------------------------------------
defaultdict을 사용시 dict의 기본값을 설정할수있음
dict,list는 매개변수로보낼때 얕은복사 또는 깊은복사 해야함(하지않을 경우 메모리를 공유하기때문)
list.copy() or copy.deepcopy(list) 깊은복사는 2차원이상 리스트일경우
'''
from collections import defaultdict
import copy
def solution(tickets):
    destination=defaultdict(list)
    for i in tickets:
        destination[i[0]].append(i[1])
        destination[i[0]].sort()
    def dfs(start,answer,dest):
        answer.append(start)
        #종료조건
        if len(answer)==1+len(tickets):
            return answer
        #재귀부분
        for i in range(len(dest[start])):
            temp=dest[start].pop(i)
            ret=dfs(temp,answer.copy(),copy.deepcopy(dest))
            dest[start].insert(i,temp)
            if ret:
                return ret
    return dfs("ICN",[],copy.deepcopy(destination))
# -----------------------------------------------------------------------------
#세건이의 코드를보고 copy를 안해도 된다는걸 깨닫고 수정함
from collections import defaultdict
def solution(tickets):
    destination=defaultdict(list)
    for i in tickets:
        destination[i[0]].append(i[1])
        destination[i[0]].sort()
    def dfs(start,answer,dest):
        answer.append(start)
        if len(answer)==1+len(tickets):
            return answer
        for i in range(len(dest[start])):
            temp=dest[start].pop(i)
            ret=dfs(temp,answer,dest)
            dest[start].insert(i,temp)
            if ret:
                return ret
        answer.pop() #추가된부분
    return dfs("ICN",[],destination)

File: Programmers/SeungDo/12주차 (안지호)/여행경로.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

bool visited[100001];
vector <string> answer;
// 알파벳이 순서가 빠른 기준으로 정렬을 한 후 DFS를 이용해 모든 경로를 구한다.
// 블로그를 참조해 풀이했습니다. https://school.programmers.co.kr/learn/courses/30/lessons/43164

 // 한번 경로를 찾았으면 더 이상 찾지 않기 위해 함수를 bool로 선언
bool dfs(int l, string start, vector<vector<string>> tickets) {
    //모든 경로를 탐색했을 경우
    if (l == tickets.size()) {
        return true;
    }

    for (int i = 0; i < tickets.size(); i++) {
        if (start == tickets[i][0] && !visited[i]) { //start가 갈 수 있는 곳 중 방문하지 않은 경로
            visited[i] = 1;
            answer.push_back(tickets[i][1]); // 경로 추가
            bool check = dfs(l + 1, tickets[i][1], tickets);
            if (check) return true; // 한번 경로를 찾았으면 더 이상 찾지 않음
            visited[i] = 0;
            answer.pop_back(); // 경로가 끊겨 있는 경우에 백트래킹
        }
    }

    return false;
}

vector<string> solution(vector<vector<string>> tickets) {
    sort(tickets.begin(), tickets.end()); // 알파벳 순서가 앞서는 경로로 정렬     

    //인천부터 시작
    answer.push_back("ICN");
    dfs(0, "ICN", tickets);

    return answer;
}
